### DISCUSSION TITLE
Crazyflie parameters

### DISCUSSION THREAD

**Discussion by uhvardhan** [bitcraze/discussions#41] [Q&A]
Hello,

I am trying to simulate the following controller [1](https://arxiv.org/pdf/2201.10509.pdf) - Page 7 Table 1 on Webots. For that I need to exact parameters for
- m = 0.027 kg (mass of the quadrotor)
- g = 9.81 m/s^2 (gravity)
- l = 0.042 m (length of arm)
- Ix=0.000016572 kg m^2 (moment of inertia around x-axis)
- Iy=0.000016656 kg m^2 (moment of inertia around y-axis)
- Iz=0.000029262 kg m^2 (moment of inertia around z-axis)
- Ir=8.06428e-05 kg m^2 (rotor drag coefficient)
- b=4.38e-05 N s^2/rad^2 (thrust coefficient)
- k = 1e-06 N s^2/rad^2 (torque coefficient)

Are these parameters right? @knmcguire 

Thank you

**Comment by knmcguire** [comment]
Good question! I've been told several times that the bachelor thesis from ETH zurich has the best system ID on the crazyflie: System Identification of the Crazyflie 2.0 Nano Quadrocopter: https://www.research-collection.ethz.ch/handle/20.500.11850/214143.I think this is a good place to start honestly. I only briefly scanned through it when I tried to find the right values for the webots model, but I till had to adjust the torque coefficient to make it flyable at that time, so it is not accurate that much. 

Small odd thing I noticed in your list is that the Ix and Iy are not symmetrical, which I would expect it to be? Where did you get this list originally?

**Reply by uhvardhan** [reply]
Ixx and Iyy mean the same thing as Ix and Iy right? 

I got them from the model.sdf from current repository - [Link](https://github.com/bitcraze/crazyflie-simulation/blob/main/gazebo-ignition/crazyflie/model.sdf)

**Reply by uhvardhan** [reply]
Thank you @knmcguire for the reply. 

I did read the bachelor thesis you mentioned. I have been playing around with k and b. The rest of the values seem right to you? Because I think you have a slightly different arm length (l) and mass (m). 

I also got Ir from your model.sdf files.

**Reply by uhvardhan** [reply]
Looks like detailed identification the parameters in itself could be a very useful study in itself.

**Reply by knmcguire** [reply]
Ahhh okay so you looked at that model.sdf? I kind of calculated that quickly to be honest... indeed the Ixx and Iyy should have been the same so don't know what went wrong there. Also the gazebo ignition model (or gazebo 2... they recently changed it) I haven't really gotten in flying condition yet. I made an issue here if anybody could get this to work with the embedded lee controller while I focus on Webots (#6 )

You should probably play around with the [crazyflie_world.wbt](https://github.com/bitcraze/crazyflie-simulation/blob/main/webots/worlds/crazyflie_world.wbt) with the crazyflie robot object in the Webots editor itself, as that is a different model, with different variables as they use a different fluid model. Also again, I haven't really looked at it very closely once I implemented it and I'm not sure if the moment coefficient is very correct.

**Reply by uhvardhan** [reply]
Yes. I looked at the model.sdf. I would love to contribute to this repository more - not just for a single crazyflie but multiple crazyflies as I told you earlier in crazyswarm discussions we had. Maybe once I finish implementing the controller I mentioned about and currently working on, I can work on here during the summer. 

Does the crazyflie_world.wbt file have the coefficients in them? I seem to have missed them. 

I was breezing through the bachelor's thesis. I may have found some of the parameters required. Thank you.

**Reply by knmcguire** [reply]
In the raw wbt file you've gotta look for [the propeller tag](https://github.com/bitcraze/crazyflie-simulation/blob/51ee5170b3b03b7eb8a0cd0427fcb3caa98befd4/webots/worlds/crazyflie_world.wbt#L142-L146) and for thrust constants and torque constants. In the editor you can find it here (the minus is necessary for the rotation of the prop):
![image](https://user-images.githubusercontent.com/11061260/164177678-f103cd64-a87a-4bc8-bf29-49a9359d7987.png)

In the webots documentation you can find all about the models they use for the propeller: https://cyberbotics.com/doc/reference/propeller

Again, this might not be super correct, as I my initial purpose was 'to just get it to fly', but now that it actually flies it would be nice to have it more realistic :smile:, so let me know if you think these numbers should be

**Reply by knmcguire** [reply]
I also believe that the inertia is based on the cilinder of the boundingbox I used, but I see that under physics, we can also give it an inertiaMatrix

**Reply by uhvardhan** [reply]
Yes. Thank you. 

I saw the constants from the .wbt file but wasn't sure if I should use them. I will work on the simulation today using the above constants and post my results here. Fingers crossed.
